### PR TITLE
Code refactoring POST/PUT/PATCH to be more like GET and use internal data type

### PR DESCRIPTION
**Pull request by ruslantalpa**
So a lot has changed and it is hard to see from the diff what exactly so here is the summary.

The point of all this is to have all the paths use the internal data type in order to generate queries. This will later allow extending the capabilities of the insert/update paths.
While implementing this i also came across a way to make all the queries (the one that wraps the core query) be similar and that gave the ability to reuse some sql fragments (count,as json, as csv).

Currently only the GET/INSERT paths have been rewritten and UPDATE will come soon.
A few new capabilities for insert, ability so send bulk insert in json format also and if return representation is set, the new records will be sent back in json/csv format as requested.

If a single record is inserted, the location header is also calculated directly in the db eliminating the need to decode the response in order to generate it.

I will also add the capability for the &select parameter to shape the response for insert in the same way it does for get.
After adding the ability to return the inserted result in csv format i encountered a few corner cases which will need work, specifically the NULL values and Boolean (the cause for one test failing at the moment)

When this is done, most of the "query transform" functions will be gone, their place will be taken by "query fragments" that are mixed and matched directly in App.hs

Please start reviewing this and see if some things can be implemented in a nicer way

### CODE REVIEW

**Comment by ruslantalpa** [comment]
Please take a closer look at the few tests i modified, especially in InsertSpec, the old ones were failing and not sure why, i've rewritten them in a new way and they work now.

File: src/PostgREST/PgStructure.hs [hunk 1]
```diff
@@ -6,8 +6,9 @@
 module PostgREST.PgStructure where
```
Comment (diogob) [inline src/PostgREST/PgStructure.hs]: We should add an explicit export list here

File: src/PostgREST/QueryBuilder.hs [hunk 2]
```diff
@@ -1,3 +1,4 @@
+{-# LANGUAGE TupleSections #-}
 module PostgREST.QueryBuilder
```
Comment (diogob) [inline src/PostgREST/QueryBuilder.hs]: I believe that if we merge ```QueryBuilder``` and ```PgQuery``` we can expose less functions and better encapsulate these modules. But to better reason about this we also need an explicit import list in the App module.

**Comment by diogob** [comment]
I have a suggestion that might be a bit out of the scope of this PR.
I think that if we would further refactor this code to follow some rules we would greatly simplify it.
The rules I propose are:
- The App module does not use any query fragments, it can only extract info from HTTP and send to QueryBuilder functions that return a complete statement for execution (things like asCsvF, wrapQuery would be forbidden).
- All SQL must be confined to one single module (hence my suggestion of merging PgQuery and QueryBuilder)

What do you think?

**Comment by ruslantalpa** [comment]
Status update:
I am reasonably happy with the way queries are constructed and the path the code takes for each request. I would suggest merging this PR so that others (probably @begriffs @diogob) can join in and start cleaning up the code.

For now i commented the PUT path, i tried implementing it but the query needed for that request was quite different from the others so it required additional parameters, like primary keys, list of columns, for a lot of functions (which already had quite ugly signatures).  Once the functions stabilise we might try adding it back.

The things to do might be:
* Rename small functions/variables to make them more descriptive
* Combine QueryBuilder and PgQuery into a single file and move some of the functions from App there
* Refactor the big ugly functions parseRequest, createStatement (by ugly i mean the signature)
* Revert back to the old way for / and /OPTIONS and get rid of some parts in dbStructure and to perform optimally when there is 1 user = 1 role situation
* There is some code duplication in App.app and App.parseRequest in the "where" section
* Rename some of the data types, in particular ApiRequest,ApiNode since it's that name is misleading, the tree structure is currently used only for the Select part of the request.
* Refactor the type of buildQuery, it's not receiving ApiRequests but structures representing queries that form an api request. Maybe it should receive a "Query" that is (Select (Tree {...}) | Update {...} | Insert {...} )

File: src/PostgREST/App.hs [hunk 3]
```diff
@@ -107,65 +84,88 @@ app dbstructure conf reqBody req =
...
-                        else M.filterWithKey (const . (`elem` pKeys)) obj
-                let params = urlEncodeVars
-                      $ map (\t -> (cs $ fst t, cs (paramFilter $ snd t)))
-                      $ sortBy (comparing fst) $ M.toList primaries
-                responseLBS status201
-                  [ jsonH
-                  , (hLocation, "/" <> cs table <> "?" <> cs params)
-                  ] $ if echoRequested then encode obj else ""
-          return $ multipart status201 responses
+      let echoRequested = hasPrefer "return=representation"
```
Comment (begriffs) [inline src/PostgREST/App.hs]: The logic to set `echoRequested` is repeated in POSt and PATCH. How about moving it to the where clause of `app`?

Also there is repetition with parsing the request and checking if the parse is a `Left` value. I realize that POST is a little different from the other verbs in that parsing the request uses `fakeSourceRelations`. But maybe there is a way to unify the construction and checking of `request` in one place.

Can you explain what the fake source relations are?

File: src/PostgREST/App.hs [hunk 4]
```diff
@@ -107,65 +84,88 @@ app dbstructure conf reqBody req =
...
+            [
+              contentTypeH,
+              (hLocation, "/" <> cs table <> "?" <> cs (fromMaybe "" location))
+            ]
+            $ if echoRequested then fromMaybe "[]" body else ""
+      where
+        request = parseRequest schema (fakeSourceRelations ++ allRels) table req reqBody
+        fakeSourceRelations = mapMaybe (toSourceRelation table) allRels
+
+    -- ([table], "PUT") ->
```
Comment (begriffs) [inline src/PostgREST/App.hs]: Might as well remove this commented block.

File: src/PostgREST/App.hs [hunk 5]
```diff
@@ -107,65 +84,88 @@ app dbstructure conf reqBody req =
...
-                responseLBS status201
-                  [ jsonH
-                  , (hLocation, "/" <> cs table <> "?" <> cs params)
-                  ] $ if echoRequested then encode obj else ""
-          return $ multipart status201 responses
+      let echoRequested = hasPrefer "return=representation"
+      case request of
+        Left e -> return $ responseLBS status400 [jsonH] $ cs e
+        Right (selectQuery, mutateQuery, isSingle) -> do
+          let pKeys = map pkName $ filter (filterPk schema table) allPrKeys -- would it be ok to move primary key detection in the query itself?
```
Comment (begriffs) [inline src/PostgREST/App.hs]: Meaning have the query select the primary keys into another column of the results as an array? Or adding something to the query that causes an error of the keys are not present?

Is your concern performance or code readability?

File: src/PostgREST/App.hs [hunk 6]
```diff
@@ -319,48 +269,137 @@ contentTypeForAccept accept
...
-      where
-        jErr = encode . object $
-          [("message", String $ "Failed to parse JSON payload. " <> cs err)]
-    Right (Object o) -> handler o
-    Right _ ->
-      return $ responseLBS status400 [jsonH] jErr
-      where
-        jErr = encode . object $
-          [("message", String "Expecting a JSON object")]
+-- handleJsonObj :: BL.ByteString -> (Object -> H.Tx P.Postgres s Response)
```
Comment (begriffs) [inline src/PostgREST/App.hs]: :skull: death to comment block!

File: test/Feature/InsertSpec.hs [hunk 7]
```diff
@@ -19,16 +19,38 @@ import TestTypes(IncPK(..), CompoundPK(..))
...
+          [json| {
+            "integer": 13, "double": 3.14159, "varchar": "testing!"
+          , "boolean": false, "date": "1900-01-01", "money": "$3.99"
+          , "enum": "foo"
+          } |]
+        liftIO $ do
+          simpleBody p `shouldBe` ""
+          simpleStatus p `shouldBe` created201
+
+      it "filters columns in result using &select" $
```
Comment (begriffs) [inline test/Feature/InsertSpec.hs]: Such a sick feature!

File: src/PostgREST/App.hs (hunk 3 above)
Comment (ruslantalpa) [inline src/PostgREST/App.hs]: When constructing the select we need relations, in the case of GET we start from a "real" table and its relations, but in case of POST/PATCH we start from the records retuned by insert/update statements which are wrapped by a "with" statement, and they act as the "fake root table" so we need to add it's relations to the allRels list and the relations are the same as the ones of the table where the mutation is happening

File: src/PostgREST/App.hs (hunk 4 above)
Comment (ruslantalpa) [inline src/PostgREST/App.hs]: I can, but while refactoring i like keeping the old code commented and close by to guide the development. In this specific case i'd like to keep it just to force me to think about adding this feature back if possible and not to leave it out just because it seemed hard at one moment

File: src/PostgREST/App.hs (hunk 5 above)
Comment (ruslantalpa) [inline src/PostgREST/App.hs]: My concern is performance, the primary keys here are only needed to calculate the location header. I wanted to move it to the query just for the fact to have all the paths be similar in behaviour, executing a single query, but i guess it's not a good idea doing introspection at every mutation request

File: test/Feature/InsertSpec.hs [hunk 8]
```diff
@@ -92,55 +114,99 @@ spec = afterAll_ resetDb $ around withApp $ do
...
+        request methodPost "/json"
+                     [("Prefer", "return=representation")]
+                     inserted
+          `shouldRespondWith` ResponseMatcher {
+            matchBody    = Just inserted
+          , matchStatus  = 201
+          , matchHeaders = ["Location" <:> [str|/json?data=eq.{"foo":"bar"}|]]
+          }
+
+        -- TODO! the test above seems right, why was the one below working before and not now
```
Comment (begriffs) [inline test/Feature/InsertSpec.hs]: I think the test fails because the json data in the Location header is actually not url encoded. Testing it in my Postman rest client seems to confirm it (unless the client UI is decoding for my benefit)

File: src/PostgREST/App.hs (hunk 6 above)
Comment (ruslantalpa) [inline src/PostgREST/App.hs]: ok

File: test/Feature/InsertSpec.hs (hunk 8 above)
Comment (ruslantalpa) [inline test/Feature/InsertSpec.hs]: so you are saying i should do "url encode" of the location header value before sending it? Like each part after the eq. ? That sucks a bit :) i need a url encoding function in sql or to return location data as an array or something and do the url encoding in haskell.

Comment (begriffs) [inline test/Feature/InsertSpec.hs]: I guess what matters is that the url in the location header is valid for later GET requests. Actually it seems like neither the raw value nor the encoded value work.

Both of these urls `http://localhost:3000/json?data=eq.%7B%22foo%22%3A%22bar%22%7D`, `http://localhost:3000/json?data=eq.{"foo":"bar"}` give me the result below.

```json
{
  "hint": "No operator matches the given name and argument type(s). You might need to add explicit type casts.",
  "details": null,
  "code": "42883",
  "message": "operator does not exist: json = unknown"
}
```

File: test/Feature/InsertSpec.hs (hunk 7 above)
Comment (diogob) [inline test/Feature/InsertSpec.hs]: Very cool indeed! :beers:

File: test/Feature/InsertSpec.hs (hunk 8 above)
Comment (begriffs) [inline test/Feature/InsertSpec.hs]: I discovered that this problem exists in 0.2.12 as well, so no need to fix it in this PR. I'll file an issue.

**Comment by begriffs** [comment]
Given the tests are all passing I'm just going to merge this into v3 then we can all contribute more easily.

File: src/PostgREST/App.hs (hunk 5 above)
Comment (diogob) [inline src/PostgREST/App.hs]: Actually, introspection is usually cheap, mutation is expensive and, in most cases, rare. So I don't see a big performance issue here.
